### ISSUE TITLE
Pipelining queries in tokio-postgres?

### ISSUE THREAD

**Issue by jprafael** [sfackler/rust-postgres#267]
Reading http://2ndquadrant.github.io/postgres/libpq-async.html suggests that the postgres protocol allows for sending multiple queries without having to wait for its responses to be received/iterated by the client. The benefit of this pipelining is that multiple RTTs are saved when the client issues multiple unrelated queries.

> A 100-statement operation run on a server 300ms round-trip-time away would take 30 seconds in network latency alone without batching; with batching it may spend as little as 0.3s waiting for results from the server.

If this is indeed possible how would it look like in terms of API? Currently the usage of `futures-state-stream` in `Connection::query()` requires consuming or discarding the stream (result rows) to get back the state (connection) incurring in lengthy RTTs.

Maybe something like a `Connection::queries()` where one gets back a sort of `StateStream` where the items themselves are streams of rows?

**Comment by sfackler** [comment]
Yep, this should totally be doable via something like the `queries` method you suggested. See https://github.com/sfackler/rust-postgres/issues/117 for a similar feature request. That was blocked on issues that should be solvable with an async implementation.

**Comment by jprafael** [comment]
Nice to know this has been proposed and will be useful to others as well!
I'll try to read up on the protocol and codease  and see if this is something simple enough that I could contribute.

**Comment by pka** [comment]
Looks like this is the missing feature that the Rust framework actix-web could also take the lead on PG centric [benchmarks](https://www.techempower.com/benchmarks/#section=test&runid=aad43f39-48a2-460c-a363-99cd543a772a&hw=ph&test=update).

**Comment by ANtlord** [comment]
I've tried `tokio-postgres` with two calls of `pg_sleep` so I see that the calls are processed consequently.

```rust
    let s1 = client.simple_query("SELECT pg_sleep(10)").for_each(|_| Ok(()))
        .map(|_| println!("first sleep is done"));
    let s2 = client.simple_query("SELECT pg_sleep(10)").for_each(|_| Ok(()))
        .map(|_| println!("second sleep is done"));
    let s1 = s1.join(s2);
    runtime.block_on(s1).unwrap();
```
So I see `first sleep is done` after 10 seconds and then `second sleep is done` after 10 seconds.
Is the feature done or do I use the API in wrong way?
tokio-postgres version is 0.4.0-rc3

**Comment by sfackler** [comment]
Pipelined calls do not execute concurrently. https://docs.rs/tokio-postgres/0.4.0-rc.3/tokio_postgres/#pipelining
